src/kubernetes.cc

```diff
@@ -453,6 +453,75 @@ KubernetesReader::GetPodAndContainerMetadata(
   return std::move(result);
 }
 
+std::vector<json::value> KubernetesReader::GetServiceList(
+    const std::string& cluster_name, const std::string& location) const
+    throw(json::Exception) {
+  std::lock_guard<std::mutex> lock(service_mutex_);
+  std::vector<json::value> service_list;
+  for (const auto& service_it : service_to_metadata_) {
+    const ServiceKey& service_key = service_it.first;
+    const std::string namespace_name = service_key.first;
+    const json::value& service_metadata = service_it.second;
+    auto endpoints_it = service_to_pods_.find(service_key);
+    const std::vector<std::string>& pod_names =
+        (endpoints_it != service_to_pods_.end()) ? endpoints_it->second
```

Comment:
This is an extension of what caused my confusion, I feel like it would be clearer to retain "endpoints_to_pods_", with a shared service key, so that this line makes sense when we read it here.

Reply:
see the other comment - the pods really belong to the service, so I prefer to keep the name. 

Hope the renaming of the local iterators helps - at this point the source matters less, we should just be referring to the caches with what they contain.